### PR TITLE
Implement deploy in execution layer

### LINKED ISSUE #805: Check `deploy_at` constructor issue
The `deploy_at` is not invoking the constructor (and thus storage writes), it needs to be fixed and tested. 

The resolution should include `deploy` operations on the execution layer instead of the transactional layer. This way the implementation will be uniform and will differ by only providing/calculating the address parameter.

### PR DESCRIPTION
**Pull request by Arcticae**
Closes #805 

## Checklist

- [ ] Linked relevant issue
- [ ] Updated relevant documentation
- [ ] Added relevant tests
- [ ] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cheatnet/tests/starknet/resources.rs [hunk 1]
```diff
@@ -6,9 +7,11 @@ use crate::common::{
 };
 use cairo_felt::Felt252;
 use cheatnet::cheatcodes::deploy::deploy;
-use cheatnet::rpc::{call_contract, CallContractOutput, ResourceReport};
+use cheatnet::rpc::{call_contract, ResourceReport};
 use conversions::StarknetConversions;
+use num_traits::ToPrimitive;
 
+// TODO (834): Verify values in this test
```
Comment (Arcticae) [inline crates/cheatnet/tests/starknet/resources.rs:14]: Tagging #834 for visibility

File: crates/cheatnet/tests/cheatcodes/deploy.rs [hunk 2]
```diff
@@ -315,3 +315,37 @@ fn deploy_invokes_constructor() {
 
     assert_success!(output, vec![Felt252::from(123)]);
 }
+
+#[test]
+fn deploy_at_invokes_constructor() {
```
Comment (MaksymilianDemitraszek) [inline crates/cheatnet/tests/cheatcodes/deploy.rs:320]: Is there any way to handle panic in constructor? If yes please add a test for asserting it.

**Review by MaksymilianDemitraszek — APPROVED** [review]
LGTM but please add requested test beforehand

File: crates/cheatnet/tests/cheatcodes/deploy.rs (hunk 2 above)
Comment (Arcticae) [reply crates/cheatnet/tests/cheatcodes/deploy.rs:320]: [Add a test case checking for the correct constructor error handling f…](https://github.com/foundry-rs/starknet-foundry/pull/828/commits/18a416ded39029313939bf492306d047d93d11cd)

File: crates/cheatnet/src/cheatcodes/deploy.rs [hunk 3]
```diff
@@ -52,72 +48,56 @@ pub fn deploy_at(
...
     );
 
-    let call_result = call_contract(
-        blockifier_state,
+    let ctor_context = ConstructorContext {
+        class_hash: *class_hash,
+        code_address: Some(contract_address),
+        storage_address: contract_address,
+        caller_address: Felt252::from(0x0000_1724_9872_3497_3219_3472_1083_7402_i128)
+            .to_contract_address(), // TODO: Extract to consts
```
Comment (war-in) [inline crates/cheatnet/src/cheatcodes/deploy.rs:62]: 👀

Comment (war-in) [reply crates/cheatnet/src/cheatcodes/deploy.rs:62]: Isn't it just the `TEST_ACCOUNT_CONTRACT_ADDRESS`?

File: crates/cheatnet/src/cheatcodes/precalculate_address.rs [hunk 4]
```diff
@@ -17,11 +14,12 @@ impl CheatnetState {
         class_hash: &ClassHash,
         calldata: &[Felt252],
     ) -> ContractAddress {
-        let account_address = ContractAddress(patricia_key!(TEST_ACCOUNT_CONTRACT_ADDRESS));
         let salt = self.get_salt();
 
         let execute_calldata = create_execute_calldata(calldata);
-        calculate_contract_address(salt, *class_hash, &execute_calldata, account_address).unwrap()
+        let deployer_address =
+            Felt252::from(0x0000_1724_9872_3497_3219_3472_1083_7402_i128).to_contract_address();
```
Comment (war-in) [inline crates/cheatnet/src/cheatcodes/precalculate_address.rs:21]: I believe it also should be extracted as a const

File: crates/cheatnet/src/cheatcodes/deploy.rs (hunk 3 above)
Comment (Arcticae) [reply crates/cheatnet/src/cheatcodes/deploy.rs:62]: No - we don't use that anymore for anything actually i believe, but i will have to verify that statement

Comment (Arcticae) [reply crates/cheatnet/src/cheatcodes/deploy.rs]: https://github.com/foundry-rs/starknet-foundry/issues/838

File: crates/cheatnet/src/cheatcodes/precalculate_address.rs (hunk 4 above)
Comment (Arcticae) [reply crates/cheatnet/src/cheatcodes/precalculate_address.rs]: done

File: crates/forge/tests/data/contracts/panicking_constructor.cairo [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    struct Storage {}
+
+    #[constructor]
+    fn constructor(ref self: ContractState) {
+        let mut panic_data = ArrayTrait::new();
+        panic_data.append('PANIK');
+        panic_data.append('DEJTA');
+        panic(panic_data);
+    }
+}
```
Comment (piotmag769) [inline crates/forge/tests/data/contracts/panicking_constructor.cairo:15]: newline

File: crates/forge/src/test_execution_syscall_handler.rs [hunk 6]
```diff
@@ -596,12 +600,12 @@ impl TestExecutionSyscallHandler<'_> {
...
 
 fn handle_deploy_result(
-    deploy_result: Result<DeployPayload, CheatcodeError>,
+    deploy_result: Result<DeployCallPayload, CheatcodeError>,
     buffer: &mut MemBuffer,
 ) -> Result<(), EnhancedHintError> {
     match deploy_result {
-        Ok(payload) => {
-            let felt_contract_address: Felt252 = payload.contract_address.to_felt252();
+        Ok(call_payload) => {
```
Comment (piotmag769) [inline crates/forge/src/test_execution_syscall_handler.rs:607]: `deploy_payload`, no?

File: crates/cheatnet/tests/cheatcodes/deploy.rs [hunk 7]
```diff
@@ -280,11 +283,11 @@ fn deploy_invalid_class_hash() {
...
 
-    assert!(match output {
-        Err(CheatcodeError::Unrecoverable(EnhancedHintError::Hint(HintError::CustomHint(msg)))) =>
-            msg.as_ref().contains(class_hash.to_string().as_str()),
-        _ => false,
-    });
+    assert!(matches!(
+        output,
+        Err(CheatcodeError::Unrecoverable(EnhancedHintError::Hint(HintError::CustomHint(msg))))
+        if msg.as_ref().contains(class_hash.to_string().as_str()),
```
Comment (piotmag769) [inline crates/cheatnet/tests/cheatcodes/deploy.rs:289]: 😮

File: crates/cheatnet/src/constants.rs [hunk 8]
```diff
@@ -52,6 +52,8 @@ pub const TEST_CONTRACT_CLASS_HASH: &str = "0x117";
 pub const TEST_ADDRESS: &str = "0x01724987234973219347210837402";
 
 pub const STEP_RESOURCE_COST: f64 = 0.01_f64;
+// snforge_std/src/cheatcodes.cairo::TEST
+pub const TEST_MEMORY_ADDRESS: &str = "0x01724987234973219347210837402";
```
Comment (piotmag769) [inline crates/cheatnet/src/constants.rs:56]: We have it defined about as `TEST_ADDRESS` xD

**Review by piotmag769 — APPROVED** [review]

File: crates/cheatnet/tests/cheatcodes/deploy.rs (hunk 7 above)
Comment (Arcticae) [reply crates/cheatnet/tests/cheatcodes/deploy.rs:289]: yeah, i know xD

File: crates/cheatnet/src/constants.rs (hunk 8 above)
Comment (Arcticae) [reply crates/cheatnet/src/constants.rs:56]: I moved the definition to `cheatnet`

**Review by piotmag769 — APPROVED** [review]

**Pull request merged**